### PR TITLE
Support for converting OpenX datasets from RLDS format to LeRobotDataset

### PR DESCRIPTION
**Pull request by michel-aractingi**
## What this does
This PR adds support to convert [Open X](https://robotics-transformer-x.github.io/) datasets in [RLDS](https://github.com/google-research/rlds) format to the dataset format of LeRobot. Full list of Open X datasets is available [here](https://docs.google.com/spreadsheets/d/1rPBD77tk60AEIGZrGSODwyyzs5FgCU9Uz3h-3_t2A9g/edit?gid=0#gid=0&range=S:S). This PR is based on a previous PR proposed in #286 . 

Open X contains 60+ datasets. Each dataset can vary in terms of the frequency, number of cameras, types of states and the names of the key/value pairs that defines their elements. The config file ```lerobot/common/datasets/push_dataset_to_hub/openx/configs.yaml``` contains the image and state keys and defines the fps so that they can be automatically detected in ```lerobot/common/datasets/push_dataset_to_hub/openx/openx_rlds_format.py``` where the ```from_raw_to_lerobot_format``` function is defined.

Transforms from the script ```lerobot/common/datasets/push_dataset_to_hub/oxe/transforms.py``` is applied to add some standardization to some the datasets.

### Download raw with gsutil tool
```
dataset=nyu_rot_dataset
gsutil -m cp -r gs://gresearch/robotics/$dataset_converted_externally_to_rlds DATA_DIR
mv DATA_DIR/$dataset_converted_externally_to_rlds/0.1.0 DATA_DIR/$dataset
```
### Download from the raw hub
```
dataset=nyu_rot_dataset
mkdir -p DATA_DIR/$dataset
python lerobot/common/datasets/push_dataset_to_hub/_download_raw.py \
                                 --raw-dir DATA_DIR/$dataset \
                                 --repo-id lerobot-raw/$dataset_raw
```
### Convert to LeRobot format and upload to your hub
```
### convert and upload
$ python lerobot/scripts/push_dataset_to_hub.py 
            --raw-dir DATA_DIR/$dataset/  
            --raw-format openx_rlds.$dataset
            --repo-id your_hub/$dataset
            --cache-dir /tmp/$dataset
```

## How to test
Test by visualizing the datasets, example:
```
datasets=(    
    "lerobot/nyu_rot_dataset"
    "lerobot/imperialcollege_sawyer_wrist_cam"
    "lerobot/utokyo_xarm_bimanual"
    "lerobot/tokyo_u_lsmo"
    "lerobot/utokyo_pr2_opening_fridge"
    "lerobot/cmu_franka_exploration_dataset"
    "lerobot/cmu_stretch"
    "lerobot/nyu_franka_play_dataset"
)
for dataset in "${datasets[@]}"; do  
 python lerobot/scripts/visualize_dataset.py --repo-id ${dataset} --episode-index 0 
done
```

Most of the datasets have been converted and uploaded to the [hub](https://huggingface.co/lerobot).

## TODO (in future PRs)
- [ ] **Depth images**: Add the observations from the depth cameras for the datasets that contain depth images. 
- [ ] **Visualize language instructions**: with the videos in `visualize_dataset_html.py` 
- [ ] **Language instructions**: Process language instructions to language embedding via tokenizers when loading the datasets.
- [ ] **Shard huge datasets**: Add sharding logic to push and load open x datasets in LeRobot format where the number of videos exceeds 10K. (around 10 datasets).

### CODE REVIEW

File: lerobot/common/datasets/push_dataset_to_hub/oxe/transforms.py
Comment (Cadene) [inline lerobot/common/datasets/push_dataset_to_hub/oxe/transforms.py:854]: Really cool. Impressive work!

File: lerobot/scripts/push_dataset_to_hub.py
Comment (Cadene) [inline lerobot/scripts/push_dataset_to_hub.py:349]: Nice

File: lerobot/scripts/push_dataset_to_hub.py
Comment (Cadene) [inline lerobot/scripts/push_dataset_to_hub.py:214]: What do you think of this simplifcation?
Could you update the comments? It's not clear to me what's the use case for `oxe_rlds` versus `oxe_rlds.bridge_orig`

```suggestion
    fmt_kwgs = {
        "raw_dir": raw_dir,
        "videos_dir": videos_dir,
        "fps": fps,
        "video": video,
        "episodes": episodes,
        "encoding": encoding,
    }

    if "oxe_rlds." in raw_format:
        # Support for official OXE dataset name inside `raw_format`.
        # For instance, `raw_format="oxe_rlds"` uses the default formating (TODO what does that mean?),
        # and `raw_format="oxe_rlds.bridge_orig"` uses the brdige_orig formating
        _, oxe_dataset_name = raw_format.split(".")
        print(f"Converting dataset [{oxe_dataset_name}] from 'oxe_rlds' to LeRobot format.")
        fmt_kwgs["oxe_dataset_name"] = oxe_dataset_name
    
    hf_dataset, episode_data_index, info = from_raw_to_lerobot_format(**fmt_kwgs)
```

File: lerobot/common/datasets/utils.py
Comment (Cadene) [inline lerobot/common/datasets/utils.py:83]: Could you add some comments on what this is doing (provides examples?) and explain why we use ` AutoTokenizer.from_pretrained(lang_tokenizer_name)`?

What are the supported `lang_tokenizer_name` ?

Why do we use `t5-small`?

File: lerobot/common/datasets/push_dataset_to_hub/oxe_rlds_format.py
Comment (Cadene) [inline lerobot/common/datasets/push_dataset_to_hub/oxe_rlds_format.py:371]: Should we move it to unit tests folder?

File: lerobot/common/datasets/push_dataset_to_hub/oxe_rlds_format.py
Comment (Cadene) [inline lerobot/common/datasets/push_dataset_to_hub/oxe_rlds_format.py:335]: Why 5 and not 30?
```suggestion
        fps = 30
```

File: lerobot/common/datasets/push_dataset_to_hub/oxe_rlds_format.py
Comment (Cadene) [inline lerobot/common/datasets/push_dataset_to_hub/oxe_rlds_format.py:273]: Example of an automatic formating that went wrong. Use `git add -p .` after automatic formatting to review the changes.
```suggestion
        path_ep_dict = tmp_ep_dicts_dir.joinpath("ep_dict_" + "0" * (10 - len(str(ep_idx))) + str(ep_idx) + ".pt")
        torch.save(ep_dict, path_ep_dict)
```

Also could you add a command on why we checkpoint?

File: lerobot/common/datasets/push_dataset_to_hub/oxe_rlds_format.py
Comment (Cadene) [inline lerobot/common/datasets/push_dataset_to_hub/oxe_rlds_format.py:28]: Nice!

**Review by Cadene — COMMENTED** [review]
Thanks for this important PR. Left a few comments. Almost ready to merge.

Why do you use `oxe` and not `openx`? I think people use `openx` more than `oxe`, also `openx` is easier to pronounce ^^ If you are ok, could rename to `openx` everywhere?

Could you save "representative" openx datasets as gitlfs artifacts for testing? See this file that you will need to modify:
```bash
python tests/scripts/save_dataset_to_safetensors.py
```
Could you add argument parser support to it?

Could you add them to `tests/test_datasets.py` in `test_backward_compatibility` function arguments?

Could you then add all your datasets to `lerobot/__init__.py` in `available_real_world_datasets`?

How did you make sure that we can visualize and train on your datasets?

Thanks!

File: lerobot/common/datasets/push_dataset_to_hub/_download_raw.py
Comment (michel-aractingi) [reply lerobot/common/datasets/push_dataset_to_hub/_download_raw.py:72]: Thanks for noticing, I am not sure how they got removed during rebasing. I added them back.

File: lerobot/scripts/push_dataset_to_hub.py
Comment (michel-aractingi) [reply lerobot/scripts/push_dataset_to_hub.py:214]: Very neat! Much better to have only one call for from_raw_to_lerobot_format.

The idea here is to allow other datasets with rlds format that are not apart of OpenX to be loaded to LeRobotDatasets.

Do you think this is an important feature to have?

File: lerobot/common/datasets/utils.py
Comment (michel-aractingi) [reply lerobot/common/datasets/utils.py:83]: Sure I'll add more insight into this. 
This is inspired by how [OpenVLA](https://github.com/openvla/openvla/blob/317dbd9e41edaa83ddf2b33f5412b5e7ddadcd4e/prismatic/models/backbones/llm/base_llm.py#L152) and [OCTO](https://github.com/octo-models/octo/blob/241fb3514b7c40957a86d869fecb7c7fc353f540/octo/data/utils/text_processing.py#L34) process the language data and converts them to language embedding since most OpenX datasets contain a ```language_instruction``` item. 

```t5-small``` is a flexible text-to-text model and the smallest of the T5 models in terms of number of params.

I am not an expert at all so if you have any suggestions regarding this part or if you want to leave it to a future PR, please let me know.

File: lerobot/common/datasets/push_dataset_to_hub/oxe/configs.yaml
Comment (michel-aractingi) [reply lerobot/common/datasets/push_dataset_to_hub/oxe/configs.yaml:4]: It comes originally from the code of @youliangtan. It is taken from the [OpenVLA](https://github.com/openvla/openvla/blob/317dbd9e41edaa83ddf2b33f5412b5e7ddadcd4e/prismatic/vla/datasets/rlds/oxe/configs.py#L33) code. It doesn't have any bearing at all on converting or loading the datasets. 

I have considered removing it many times. It acts as an extra bit of information for understanding how state and action spaces are constructed.

File: lerobot/common/datasets/push_dataset_to_hub/oxe/configs.yaml
Comment (michel-aractingi) [reply lerobot/common/datasets/push_dataset_to_hub/oxe/configs.yaml:14]: This only concerns the very few navigation datasets in OpenX. The action space is not unified between them, for some [papers](https://arxiv.org/abs/1709.10489) it the steering angle for [others](https://arxiv.org/pdf/2104.05859) its more complicated.

File: lerobot/common/datasets/push_dataset_to_hub/oxe_rlds_format.py
Comment (michel-aractingi) [reply lerobot/common/datasets/push_dataset_to_hub/oxe_rlds_format.py:273]: I am using this checkpointing for very big datasets in case they crash. I will remove it since there is no need to have that in the main code.

**Comment by Cadene** [comment]
How could we visualize the language instructions in `visualize_dataset_html.py` ?

File: lerobot/scripts/push_dataset_to_hub.py
Comment (Cadene) [inline lerobot/scripts/push_dataset_to_hub.py:6]:
```suggestion
# you may not use this file except in compliance with the License.
```

**Comment by Cadene** [comment]
For a future PR, we should structure [`self.info`](https://github.com/huggingface/lerobot/blob/main/lerobot/common/datasets/lerobot_dataset.py#L69) of the dataset.

For instance, we should specify:
- robot name (koch, aloha, etc.)
- type of robots (manipulation, bimanual manipulation, mobile manipulation, navigation, etc.)
- type of tasks (pick and place, multitask, etc.)
- names of each state dimension, names of each action dimension
- type of state and action spaces:
  - end effector euler position
  - end effector quaternion
  - end effector velocity
  - joint position 
  - joint velocity (??)
  - mixed (??)

cc @aliberts @alexander-soare

File: lerobot/common/datasets/utils.py
Comment (youliangtan) [reply lerobot/common/datasets/utils.py:83]: This was introduced for language conditioned task. Adding `lang_tokenizer_name` is to support different language encoder.

File: lerobot/common/datasets/push_dataset_to_hub/oxe/configs.yaml
Comment (youliangtan) [reply lerobot/common/datasets/push_dataset_to_hub/oxe/configs.yaml]: This config file is adapted from Octo (OpenVLA doesnt use propio), which we ensure all different types of state encoding have the same number of inputs, thus the padding in `POS_EULER`

File: lerobot/common/datasets/push_dataset_to_hub/oxe/configs.yaml
Comment (youliangtan) [reply lerobot/common/datasets/push_dataset_to_hub/oxe/configs.yaml]: This `JOINT_BIMANUAL` is intended to be used only for the table-top aloha (not used in this conversion config). We can remove/comment this line too to prevent confusion too.

**Comment by youliangtan** [comment]
Thanks @michel-aractingi for helping improve and convert the remaining dataset according to my previous PR. :+1: 

Regarding the naming of `openx`, I think we normally call it `oxe`, which it is more explicit for "embodiments".

**Comment by michel-aractingi** [comment]
Thanks @Cadene  for your comprehensive review. 

> Why do you use oxe and not openx? 

I agree I changed the name from oxe to openx. 

> Could you save "representative" openx datasets as gitlfs artifacts for testing? 

Done I saved nyu_franka_play_dataset and berkeley_cable_routing as git lfs artifiacts and added them to the ```tests/test_datasets.py```.

> Could you add argument parser support to it?

I am not sure I understand. Should I add argument parse to ```tests/scripts/save_datasets_to_safetensors.py```

> Could you then add all your datasets to lerobot/__init__.py in available_real_world_datasets?

Done .

> How did you make sure that we can visualize and train on your datasets?

I was testing with ```lerobot/scripts/visualize_datasets.py```. I am not sure how to test for training. 

example:
```
datasets=(    
    "lerobot/nyu_rot_dataset"
    "lerobot/imperialcollege_sawyer_wrist_cam"
    "lerobot/utokyo_xarm_bimanual"
    "lerobot/tokyo_u_lsmo"
    "lerobot/utokyo_pr2_opening_fridge"
    "lerobot/cmu_franka_exploration_dataset"
    "lerobot/cmu_stretch"
    "lerobot/nyu_franka_play_dataset"
)
for dataset in "${datasets[@]}"; do  
 python lerobot/scripts/visualize_dataset.py --repo-id ${dataset} --episode-index 0 
done
```

File: lerobot/common/datasets/push_dataset_to_hub/oxe_rlds_format.py
Comment (aliberts) [reply lerobot/common/datasets/push_dataset_to_hub/oxe_rlds_format.py]: I wouldn't mind if it stays. I also had issues with UMI and did something similar. Since these are (hopefully) one-time use scripts, I think it's okay to have them in the state you were actually able to run them.

**Review by aliberts — COMMENTED** [review]
Thank you for porting all these datasets @michel-aractingi!

Right now your b/c tests are failing because you need to add `nyu_franka_play_dataset` and `berkeley_cable_routing` dataset samples to `tests/data/lerobot` as they're also needed to run this test. See running the push_dataset_to_hub script with the `--tests-data-dir` arg.

I'm not sure we want all of them though as this folder is already quite heavy and we're going to get rid of it soon enough. Just these 2 should be ok.

**Comment by michel-aractingi** [comment]
> Thank you for porting all these datasets @michel-aractingi!
> 
> Right now your b/c tests are failing because you need to add `nyu_franka_play_dataset` and `berkeley_cable_routing` dataset samples to `tests/data/lerobot` as they're also needed to run this test. See running the push_dataset_to_hub script with the `--tests-data-dir` arg.
> 
> I'm not sure we want all of them though as this folder is already quite heavy and we're going to get rid of it soon enough. Just these 2 should be ok.

Thanks Simon for the clarification! I added the missing files in `tests/data/lerobot` the test still fails because of the dependency on transformers for the language instructions in `lerobot/common/datasets/utils.py`. How should we address this? Do we leave it for a future PR?
cc @aliberts @Cadene

File: lerobot/common/datasets/push_dataset_to_hub/_download_raw.py
Comment (aliberts) [inline lerobot/common/datasets/push_dataset_to_hub/_download_raw.py:134]: I have 2 questions about this:
- Did you merge some raw datasets together to produce non-raw datasets? If so, is there a table somewhere I can use to know which LeRobot dataset (in `lerobot.available_datasets`) were create from which of these raw datasets? I'm seeing 62 new raw datasets here while there's only 40 new datasets in `lerobot.available_datasets`.
- If not, we should keep the naming convention we previously used: `lerobot/some_dataset` should be coming from `lerobot-raw/some_dataset_raw`. If you do this, no need to reupload everything as I believe you can simply do something like this:
```python
from huggingface_hub import HfApi

old_new_ids = {
    "lerobot-raw/some_old_name": "lerobot-raw/some_new_name",
    "lerobot-raw/some_other_old_name": "lerobot-raw/some_other_new_name",
}

api = HfApi()
for old_name, new_name in old_new_ids.items():
    api.move_repo(from_id=old_name, to_id=new_name, repo_type="dataset")
```

File: lerobot/common/datasets/utils.py
Comment (aliberts) [inline lerobot/common/datasets/utils.py:75]: Several remarks here:
- Is there any way we can avoid adding transformer as a dependency if it's just for a tokenizer, at least for now? Ok if we can't, just asking as it's quite heavy. It should at least be an optional (extra) dependency, and we should specify clearly which datasets need it so that the would user know whether he/she should install it or not.
- I don't see any sign of transformers being added to the projects' dependency, which is why it's failing on the CI. To add a dependency:
```diff
# pyproject.toml

[tool.poetry.dependencies]
+ transformers = {version = ">=4.44.1", optional = true}

[tool.poetry.extras]
+ language = ["transformers"]
```
Then run
```bash
poetry lock --no-update
```

This will make `transformers` installable with lerobot as an extra:
```bash
# using pip
pip install -e ".[language]"

# using poetry
poetry install . --extras "language"
```
- As it would be an optional dependency, it should not be imported without "safety measures", e.g.:
```python
try:
    importlib.import_module(transformers)
except ModuleNotFoundError as e:
    print(
        f"transformers is not installed. Please install it with `pip install 'lerobot[language]'`"
    )
    raise e
```
- For the CI to pass, you would need to use the same trick we use in e.g. [this test](https://github.com/huggingface/lerobot/blob/996468bcceafe6688a74d321550fced20f8ec83b/tests/test_push_dataset_to_hub.py#L265) where we use a `@require_package_arg` to specify which packages are required for the test to be run (and the test is skipped otherwise if it's not installed)

Don't hesitate to ping me or send me a DM if any of this isn't clear ;)

Comment (michel-aractingi) [reply lerobot/common/datasets/utils.py:75]: Thanks @aliberts ! it is super clear. 

But I am considering just removing it. We can leave up to the user to choose the method of how to address language conditioned tasks. 

We can just ignore the 'str' instances in `utils.hf_transform_to_torch`. What do you think?

**Review by aliberts — APPROVED** [review]
Thank you @michel-aractingi!
Before merging, could you remove `cmu_stretch` and `nyu_franka_play_dataset` from `tests/data/save_dataset_to_safetensors` and `tests/data/lerobot` as we won't be using them for now?

Everything LGTM otherwise ;)

**Comment by yunho-c** [comment]
@michel-aractingi I found this PR while trying to convert a non-`LeRobot` dataset into `LeRobotDataset`. I learned a lot reading through the code! However, it seems like modifications from this PR are reverted as of latest main branch; if you could point me to where to look instead or provide any tips for porting an already-existing dataset, I would really appreciate it! (Sorry for reviving an old thread!)

**Pull request merged**
